c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -388,25 +388,36 @@ class THCCachingAllocator {
     return basePtr;
   }
 
-  void recordStream(void* ptr, cuda::CUDAStream stream) {
+  void recordStream(const DataPtr& ptr, cuda::CUDAStream stream) {
     // Empty tensor's storage().data() might be a null ptr. As there is no
     // blocks associated with those tensors, it is fine to do nothing here.
-    if (!ptr) {
+    if (!ptr.get()) {
       return;
     }
 
+    // The ptr is not allocated by CUDA, most likely user error
+    TORCH_CHECK(
+        ptr.device().type() == kCUDA, "Tensor is not allocated in CUDA");
+
+    // If a tensor is not allocated by this instance, simply skip
+    // This usually happens when CUDA tensors are shared across processes,
+    // we have implemented reference counting based sharing mechanism to
+    // guarantee tensors won't be accidentally freed by one process while
+    // they are still being used in another
```

Comment:
While it is true that we've implemented distributed reference count, that's not actually germane to why it's OK to skip: the real reason it's OK to skip is that record stream fixes a synchronization problem that is introduced by our caching allocator; there is no such problem for cuda allocations that come from other sources.

Reply:
Not sure what synchronization problem you were referring to, but I think for the scenario I listed above, it will not work if the reference count is not implemented, so I guess it must be part of the reason. 
Will you add a bit more context as what is the synchronization problem and why fixing of that problem enables us to skip here.

Reply:
> I think for the scenario I listed above, it will not work if the reference count is not implemented

Not really. In the old world when we did not have a distributed refcount, there was an extra obligation on the part of users which was that they were obligated to keep CUDA tensors live as long as there might be another process accessing them. So once again, there is no stream safety problem at the destination side: the memory is guaranteed not to be freed back to the caching allocator and so there can't be a stream race.

The beginning of the explanation for what recordStream does is in this comment:

```
// The library provides a recordStream() function to help insert the correct
// synchronization when allocations are used on multiple streams. This will
// ensure that the block is not reused before each recorded stream completes
// work.
```

I'm not sure what more context you are looking for; if you can ask some specific questions that will help us make the description better.

Reply:
Ok, so you are saying without ref count, it is users's obligation to keep the tensor alive when it is shared across processes, I thought there was another safeguard. 
My point is, the ref count guarantees the memory of tensor won't be freed even if a user makes a mistake, that is why I think it is ok to not print any warning here, so the above explanation still stands.